src/sinks/databend/service.rs

```diff
@@ -0,0 +1,217 @@
+use bytes::Bytes;
+use std::collections::BTreeMap;
+use std::task::{Context, Poll};
+use std::time::SystemTime;
+
+use futures::future::BoxFuture;
+use rand::{thread_rng, Rng};
+use rand_distr::Alphanumeric;
+use snafu::Snafu;
+use tower::Service;
+use vector_common::finalization::{EventFinalizers, EventStatus, Finalizable};
+use vector_common::internal_event::CountByteSize;
+use vector_common::request_metadata::{MetaDescriptive, RequestMetadata};
+use vector_core::stream::DriverResponse;
+
+use crate::{internal_events::EndpointBytesSent, sinks::util::retries::RetryLogic};
+
+use super::{
+    api::{DatabendAPIClient, DatabendHttpRequest, DatabendPresignedResponse},
+    error::DatabendError,
+};
+
+#[derive(Clone)]
+pub struct DatabendRetryLogic;
+
+impl RetryLogic for DatabendRetryLogic {
+    type Error = DatabendError;
+    type Response = DatabendResponse;
+
+    fn is_retriable_error(&self, error: &Self::Error) -> bool {
+        match error {
+            DatabendError::Server { code, message: _ } => match *code {
+                429 => true,
+                // general server error
+                500 => true,
+                // storage doesn't support presign operation
+                3902 => false,
+                // fail to parse stage attachment
+                1046 => false,
+                _ => false,
+            },
+            _ => false,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct DatabendService {
+    client: DatabendAPIClient,
+    database: String,
+    table: String,
+}
+
+#[derive(Clone)]
+pub(crate) struct DatabendRequest {
+    pub data: Bytes,
+    pub finalizers: EventFinalizers,
+    pub metadata: RequestMetadata,
+}
+
+impl Finalizable for DatabendRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        self.finalizers.take_finalizers()
+    }
+}
+
+impl MetaDescriptive for DatabendRequest {
+    fn get_metadata(&self) -> RequestMetadata {
+        self.metadata
+    }
+}
+
+#[derive(Debug, Snafu)]
+pub struct DatabendResponse {
+    metadata: RequestMetadata,
+}
+
+impl DriverResponse for DatabendResponse {
+    fn event_status(&self) -> EventStatus {
+        EventStatus::Delivered
+    }
+
+    fn events_sent(&self) -> CountByteSize {
+        CountByteSize(
+            self.metadata.event_count(),
+            self.metadata.events_estimated_json_encoded_byte_size(),
+        )
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.metadata.request_encoded_size())
+    }
+}
+
+impl DatabendService {
+    pub(crate) const fn new(
+        client: DatabendAPIClient,
+        database: String,
+        table: String,
+    ) -> DatabendService {
+        DatabendService {
+            client,
+            database,
+            table,
+        }
+    }
+
+    pub(crate) fn new_stage_location(&self) -> Result<String, DatabendError> {
+        let now = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH)?;
```

Comment:
```suggestion
        let now = Utc::now().timestamp();
```

I think we generally use `Utc` - is there a particular reason for `SystemTime` here?

Reply:
no particular reason, changed to `Utc`